File: examples/orders-service/src/repository/orderRepository.ts
import type { Order } from "../domain/order.js";

/** Persistence contract. Expanding it in Meridian reveals the methods and their implementations. */
export interface OrderStore {
  save(order: Order): void;
  findById(id: string): Order | undefined;
  count(): number;
}

/** Stores orders and reads them back. In-memory stand-in for a database. */
export class OrderRepository implements OrderStore {
  private readonly byId = new Map<string, Order>();

  /** Persist an order so it can be looked up later. */
  save(order: Order): void {
    this.byId.set(order.id, order);
  }

  /** Find a previously saved order, or undefined if we have never seen it. */
  findById(id: string): Order | undefined {
    return this.byId.get(id);
  }

  /** How many orders we are currently holding. */
  count(): number {
    return this.byId.size;
  }
}

File: examples/orders-service-py/orders/repository/order_repository.py
"""Stores orders and reads them back."""

from __future__ import annotations

from ..domain.order import Order


class OrderRepository:
    """In-memory stand-in for a database."""

    def __init__(self) -> None:
        self._by_id: dict[str, Order] = {}

    def save(self, order: Order) -> None:
        """Persist an order so it can be looked up later."""
        self._by_id[order.id] = order

    def find_by_id(self, order_id: str) -> Order | None:
        """Find a previously saved order, or None if we have never seen it."""
        return self._by_id.get(order_id)

    def count(self) -> int:
        """How many orders we are currently holding."""
        return len(self._by_id)
